src/submodule.c

```diff
@@ -1854,6 +1854,11 @@ static void submodule_release(git_submodule *sm)
 	git__free(sm);
 }
 
+int git_submodule_dup(git_submodule **out, git_submodule *source)
+{
+	return git_object_dup((git_object **)out, (git_object *)source);
```

Comment:
That won't work — `git_submodule` is not a proper subtype/C "subclass" of `git_object`, so this will actually clobber memory in the underlying structure when incrementing the refcount.

IIRC your previous version (using `GIT_REFCOUNT_INC`) was correct but the function's prototype wasn't.

Reply:
I can't figure out how to adjust `*out` parameter with previous implementation.
Something like this?
```swift
int git_submodule_dup(git_submodule **out, git_submodule *source)
{
  if (!out) { return -1; }
  if (!source) { return -1; }
  *out = source;
  GIT_REFCOUNT_INC(*out);
  return 0;
}
```

Reply:
By the way, project has different styles in spaces/tabs and nullability check.
Is there any (current) guide to look at?

Reply:
There's docs/coding-style.md, which is supposed to be current and seems to be. I'd maybe add something about the assertion macros (which are new-ish) and when to use them — ideally for "impossible" or "API misuse" conditions.